src/buckets.ts

```diff
@@ -0,0 +1,253 @@
+      debug: context.debug,
+    }
+  }
+
+  /**
+   * Initializes a new bucket.
+   * @param name Human-readable bucket name. It is only meant to help identify a bucket in a UI and is not unique.
+   * @param ctx Context object containing web-gRPC headers and settings.
+   */
+  async init(name: string, ctx?: Context) {
+    logger.debug('init request')
+    const req = new pb.InitRequest()
+    req.setName(name)
+    const res: pb.InitReply = await this.unary(API.Init, req, ctx)
+    return res.toObject()
+  }
+
+  /**
+   * Returns a list of all bucket roots.
+   * @param ctx Context object containing web-gRPC headers and settings.
+   */
+  async list(ctx?: Context) {
+    logger.debug('list request')
+    const req = new pb.ListRequest()
+    const res: pb.ListReply = await this.unary(API.List, req, ctx)
+    return res.toObject().rootsList
+  }
+
+  /**
+   * Returns information about a bucket path.
+   * @param key Unique (IPNS compatible) identifier key for a bucket.
+   * @param path A file/object (sub)-path within a bucket.
+   * @param ctx Context object containing web-gRPC headers and settings.
+   */
+  async listPath(key: string, path: string, ctx?: Context) {
+    logger.debug('list path request')
+    const req = new pb.ListPathRequest()
+    req.setKey(key)
+    req.setPath(path)
+    const res: pb.ListPathReply = await this.unary(API.ListPath, req, ctx)
+    return res.toObject()
+  }
+
+  /**
+   * Removes an entire bucket. Files and directories will be unpinned.
+   * @param key Unique (IPNS compatible) identifier key for a bucket.
+   * @param ctx Context object containing web-gRPC headers and settings.
+   */
+  async remove(key: string, ctx?: Context) {
+    logger.debug('remove request')
+    const req = new pb.RemoveRequest()
+    req.setKey(key)
+    await this.unary(API.Remove, req, ctx)
+    return
+  }
+
+  /**
+   * Returns information about a bucket path.
+   * @param key Unique (IPNS compatible) identifier key for a bucket.
+   * @param path A file/object (sub)-path within a bucket.
+   * @param ctx Context object containing web-gRPC headers and settings.
+   */
+  async removePath(key: string, path: string, ctx?: Context) {
+    logger.debug('remove path request')
+    const req = new pb.RemovePathRequest()
+    req.setKey(key)
+    req.setPath(path)
+    await this.unary(API.RemovePath, req, ctx)
+    return
+  }
+
+  /**
+   * Pushes a file to a bucket path.
+   * @param key Unique (IPNS compatible) identifier key for a bucket.
+   * @param path A file/object (sub)-path within a bucket.
+   * @param input The input file/stream/object.
+   * @param ctx Context object containing web-gRPC headers and settings.
+   * @param opts Options to control response stream. Currently only supports a progress function.
+   * @note This will return the resolved path and the bucket's new root path.
+   */
+  async pushPath(key: string, path: string, input: any, ctx?: Context, opts?: { progress?: (num?: number) => void }) {
```

Comment:
If you follow this along, it matches the go version quite closely. The fun difference here is, we return a promise, which resolves with the final message from the server. So we're wrapping the whole bidi dance into a single promise. It feels like a nice abstraction for the developer, so callbacks (unless you want progress) needed.

Reply:
Very nice